Add scrollbaroptions FormSpec element

v-rob opened the pull request:
This PR adds a `scrollbaroptions` element for formspecs.  Similar to `tableoptions`, it must be placed before the scrollbars it modifies.  Old scrollbars are not affected.  Closes #8575.

Valid options are:

- `min` - Minimum value of the scrollbar (default 0)
- `max` - Maximum value of the scrollbar (default 1000)
- `smallstep` - Scroll amount when using arrows or scroll wheel (default 10)
- `largestep` - Scroll amount when clicking the bar itself (default 100)
- `thumbsize` - Size of the thumb determined by the number of unit the thumb spans out of the range of values (max - min).
- `arrows` - Whether to make the arrows hidden, shown, or automatic.

Some potential examples:

`scrollbaroptions[min=0;max=255;smallstep=8;largestep=16;thumbsize=8;arrows=hide]`
`scrollbaroptions[max=10000;largestep=1000]`

If the step value is negative, it will use Irrlicht's defaults.  If the max is less than or equal to the min, the scrollbar element will have no bar and do nothing, i.e. disabled.  If thumbsize is zero or less, it defaults to 1 (which usually will appear as a square).

This PR is Ready for Review.

DS-Minetest commented:
Imo it would be good if #8507 is merged before this, so that here the option sliderlength can be added.

v-rob commented:
I'm trying to incorporate the new `guiScrollBar` class in this PR, and I can't figure out what's going incorrectly.  The [error code](https://travis-ci.org/minetest/minetest/jobs/542035206#L1154) is rather undecipherable, and I can't figure anything out from it.  I think the error lies in line [516](https://github.com/minetest/minetest/pull/8530/files#diff-074850a88796e13dad12dd7ae1a8d9f7R516), but I'm not certain.  I based this `guiScrollBar` on the one in `guiEditBoxWithScrollbar.cpp`

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -513,14 +513,23 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+			if (is_horizontal) {
+				e->setPageSize(dim.X);
+			} else {
+				e->setPageSize(dim.Y);
```
Shouldn't `dim.X` and `dim.Y` be swapped?

v-rob replied:
No, since if the scrollbar is horizontal, the page size should be the width, X.  Otherwise, it should be the height, Y.

DS-Minetest replied:
Oh, soory, I misunderstood pagesize.

DS-Minetest commented:
Have you rebased already?

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -545,25 +554,38 @@ void GUIFormSpecMenu::parseScrollBarOptions(parserData* data, const std::string
 			}
 
-			errorstream << "Invalid scrollbaroptions option(" << options[0] <<
+			errorstream << "Invalid scrollbaroptions option(" << option <<
```
For future releases: Please change this to `warningstream` or `infostream`.

v-rob commented:
I have rebased as evidenced by this file: https://github.com/v-rob/minetest/blob/scrollbaroptions/src/gui/guiScrollBar.cpp.  I believe the problem lies in the instantiation of `guiScrollBar`, but I don't have much to work off of.  Maybe @stujones11 can shed some light on this, since he made it.

DS-Minetest commented:
@v-rob `"guiScrollBar.h"` is not yet included.

rubenwardy commented:
This is really cool!

v-rob commented:
Well, it works now (almost).  I just needed to include `guiScrollBar.h` (duh!) as well as change a few `IGUIScrollBar` things to the new `guiScrollBar`.  Now I just need to make the bar change size properly -- see the screenshot :)

![large_scrollbar](https://user-images.githubusercontent.com/31123645/59151166-30c04800-89e3-11e9-99bd-f07c7d13761f.png)

v-rob commented:
I still have two bugs to iron out before this PR can be merged:  First, I _still_ don't think that I have `pagesize` quite right.  The other bug is that trying to get the value from the scrollbar always returns `INV` (which is failure).  I need to do some more testing.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -513,14 +514,30 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+		guiScrollBar* e = new guiScrollBar(Environment, this, spec.fid, rect,
+				is_horizontal, true);
+
+		int max = data->scrollBarOptions.max;
```
As far as I have seen, minetest seems to use `s32`, not `int`.

DS-Minetest commented:
If you want the default quadratic slider rectangle, you have to set page_size as large as possible, this results in a thumb_size of (nearly) 0, which is smaller than thumb_min and will ergo be set to thumb_min.

If the length of the scrollbar (including borders) is the length of the visible part of the whole page that you scroll over, page_size is the length of the whole page.
(All this lengths are in pixels.)

The length of the scrollbar over page_size is kept equal to thumb_size over thumb_area.

Imo, this is very confusing and not really useful for modders.
There should be something else that you can set.

Maybe just give the thumb size in scrollbar units.

Example 1: The whole page is 1000 scrollbar units long (max-min). (In pixels it's `1000*some_factor`.) The visible part is `10*same_factor_as_before` pixels long. => The thumb has to be 10 long (in scrollbar units).

Example 2: You can choose one number out of 10 with a scrollbar (from 0 to 9). Ergo min is 0 and max is 9. And thumb_size in scrollbar units is 1 because it's always one number at a time.

Now, the question is: How to convert thumb_size in scrollbar units into page_size in pixels?

stujones11 commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -513,14 +514,30 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
 		spec.send  = true;
-		gui::IGUIScrollBar* e =
-				Environment->addScrollBar(is_horizontal,rect,this,spec.fid);
+		guiScrollBar* e = new guiScrollBar(Environment, this, spec.fid, rect,
```
These references must be 'dropped' in the formspec destructor otherwise it will lead to memory leaks!

DS-Minetest replied:
[Destructer](https://github.com/minetest/minetest/blob/5106ea9e39a4ac34e645baae419a5ad1f2a5cb91/src/gui/guiFormSpecMenu.cpp#L119) and [regenerateGui](https://github.com/minetest/minetest/blob/5106ea9e39a4ac34e645baae419a5ad1f2a5cb91/src/gui/guiFormSpecMenu.cpp#L2057) call [`removeChildren`](https://github.com/minetest/minetest/blob/5106ea9e39a4ac34e645baae419a5ad1f2a5cb91/src/gui/guiFormSpecMenu.cpp#L159) [which drops all children](http://irrlicht.sourceforge.net/docu/_i_g_u_i_element_8h_source.html#l00304) of `this` which include `e`.

stujones11 replied:
No, the children are removed, not dropped. Elements that are 'added' with `Evironment-add..` should not be dropped.

DS-Minetest replied:
Oh, [`addChildToEnd`](http://irrlicht.sourceforge.net/docu/_i_g_u_i_element_8h_source.html#l00818) grabs the element. Well, then I think, it would be ok to drop `e` directly after constructing it.

stujones11 replied:
I suggest you look at how tables are handled `m_tables` there are also created with the `new` operator, hence why they are being 'dropped' here: https://github.com/minetest/minetest/blob/5106ea9e39a4ac34e645baae419a5ad1f2a5cb91/src/gui/guiFormSpecMenu.cpp#L122

Essentially, the author needs to do the same for `m_scrollbars`. It may be possible to drop them before hand, however, it is more usual to do this in the destructor.

DS-Minetest replied:
Maybe just drop the element directly after creating it if it is never used again directly by this class.

stujones11 replied:
Are you sure, or is this a guess? I would prefer to err on the side of caution and consistency and do it like they did it for tables :)

DS-Minetest replied:
I'm not really a trustworthy source here. Other humans are more experienced.

v-rob replied:
I think I have this done.  I just copied the table and changed it a bit, so I want someone to check it over.

DS-Minetest replied:
It looks good now.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2932,9 +3007,9 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
 					IGUIElement * element = getElementFromId(s.fid);
-					gui::IGUIScrollBar *e = NULL;
+					guiScrollBar *e = NULL;
 					if ((element) && (element->getType() == gui::EGUIET_SCROLL_BAR)) {
```
`guiScrollBar` is not of type `gui::EGUIET_SCROLL_BAR`.
It is `EGUIET_ELEMENT`:
https://github.com/minetest/minetest/blob/429a98964859b83016f2eb47a47a08ab8dc3c57e/src/gui/guiScrollBar.cpp#L19

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2932,9 +3007,9 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
+					guiScrollBar *e = NULL;
 					if ((element) && (element->getType() == gui::EGUIET_SCROLL_BAR)) {
-						e = static_cast<gui::IGUIScrollBar*>(element);
+						e = static_cast<guiScrollBar*>(element);
```
```suggestion
						e = static_cast<guiScrollBar *>(element);
```

DS-Minetest commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -405,7 +405,7 @@ class GUIFormSpecMenu : public GUIModalMenu
 	std::map<std::string, TooltipSpec> m_tooltips;
 	std::vector<std::pair<irr::core::rect<s32>, TooltipSpec>> m_tooltip_rects;
-	std::vector<std::pair<FieldSpec,gui::IGUIScrollBar*> > m_scrollbars;
+	std::vector<std::pair<FieldSpec,guiScrollBar*> > m_scrollbars;
```
```suggestion
	std::vector<std::pair<FieldSpec, guiScrollBar *> > m_scrollbars;
```

DS-Minetest commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -455,6 +455,15 @@ class GUIFormSpecMenu : public GUIModalMenu
+			int min = 0;
+			int smallStep = 10;
+			int largeStep = 100;
+			int thumbSize = 1;
```
Maybe use `s32` instead of `int`.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -529,6 +542,63 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+	std::vector<std::string> parts = split(element, ';');
+
+	if (parts.size() > 0) {
+		for (unsigned int i = 0; i < parts.size(); i++) {
```
Maybe `u32` instead of `unsigned int`.

v-rob replied:
I don't think it's necessary to use this for a counter.

DS-Minetest replied:
Actually here you could just do something like `for ( const std::string &i : parts)`.

v-rob replied:
I suppose you could.  C++ has many strange syntactical things you can do.

DS-Minetest replied:
A ranged for loop isn't that strange…

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -529,6 +542,63 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+{
+	std::vector<std::string> parts = split(element, ';');
+
+	if (parts.size() > 0) {
```
```suggestion
	if (parts.size() <= 0) {
		warningstream << ...;
		return;
	}
	...
```

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -529,6 +542,63 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+				if (value < 0)
+					data->scrollBarOptions.smallStep = 10;
+				else
+					data->scrollBarOptions.smallStep = value;
```
This would be shorter:
```cpp
data->scrollBarOptions.smallStep = MYMAX(value, 10);
```
(Also everywhere else.)

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -513,14 +514,26 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
 		spec.send  = true;
-		gui::IGUIScrollBar* e =
-				Environment->addScrollBar(is_horizontal,rect,this,spec.fid);
+		guiScrollBar* e = new guiScrollBar(Environment, this, spec.fid, rect,
```
```suggestion
		guiScrollBar *e = new guiScrollBar(Environment, this, spec.fid, rect,
```

DS-Minetest commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -508,6 +517,7 @@ class GUIFormSpecMenu : public GUIModalMenu
 	bool parseVersionDirect(const std::string &data);
 	bool parseSizeDirect(parserData* data, const std::string &element);
 	void parseScrollBar(parserData* data, const std::string &element);
+	void parseScrollBarOptions(parserData* data, const std::string &element);
```
```suggestion
	void parseScrollBarOptions(parserData *data, const std::string &element);
```

DS-Minetest replied:
This is now `parserData * data`.

DS-Minetest commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -405,7 +405,7 @@ class GUIFormSpecMenu : public GUIModalMenu
 	std::map<std::string, TooltipSpec> m_tooltips;
 	std::vector<std::pair<irr::core::rect<s32>, TooltipSpec>> m_tooltip_rects;
-	std::vector<std::pair<FieldSpec,gui::IGUIScrollBar*> > m_scrollbars;
+	std::vector<std::pair<FieldSpec,guiScrollBar*> > m_scrollbars;
```
Does this even work without including `"guiScrollBar.h"` in `guiFormSpecMenu.h`?

stujones11 replied:
I agree, either the include should be moved to the header or use a forward declaration, not sure which is preferred in MT but since `guiTable.h` is included in the header, I would go with the former.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -529,6 +542,63 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+				continue;
+			}
+			if (options[0] == "thumbsize") {
+				if (value < 0)
```
```suggestion
				if (value <= 0)
```
Else you would divide by 0.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -529,6 +545,54 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+	}
+
+	for (unsigned int i = 0; i < parts.size(); i++) {
+
```
It might be that empty lines after things like `for` are not wanted.
```suggestion
```

DS-Minetest commented:
Code looks nice now.

Can this PR be added to `Proposed order of merging formspec PRs` at https://dev.minetest.net/Meetings#2019-06-X (between style and real_coordinates)?

v-rob commented:
Yes, I think it's ready to be merged now.  It can be after or before `style`, but I think it should be before `scroll_container`, since this directly modifies scrollbars.

DS-Minetest commented:
Tested. Almost everything works nicely.

But if you set `min`, it's drawn incorrectly.
Maybe this can be fixed by using `scroll_pos-min` instead of `scroll_pos` here:
https://github.com/minetest/minetest/blob/bb14b7e9564b982ddcecb972218bfdae00cb4de4/src/gui/guiScrollBar.cpp#L292

v-rob commented:
Rebased and fixed improper scrollbar drawing.  Travis CI complains about this line, although it adheres to the Minetest code conventions: https://travis-ci.org/minetest/minetest/jobs/555143252#L522

DS-Minetest commented:
Tested, it works pretty well.
However, if I set `min` and `max` to nearly equal values (eg. `0` and `3`), I can't get the scroller to be quadratic anymore.

(I would just add the extra tab to please travis.)

Your rebase is a little weird. The reviewing diff shows all your PR's changes and also the changes of recent master commits. Reviewing won't be easily possible like this.

SmallJoker commented:
Something went wrong in the rebase. Not all of your commits were applied to the top of your branch.

v-rob commented:
I don't know why the rebase did what it did, and I have no idea how to fix it.  Git is mostly black magic to me still.  I would appreciate any help I could get on this.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2180,6 +2243,11 @@ void GUIFormSpecMenu::parseElement(parserData* data, const std::string &element)
 		return;
 	}
 
+	if (type == "scrollbaroptions") {
```
Move below "real_coodinates".

SmallJoker commented:
Rebased & squashed a few similar commits. This PR would fit very well into `style[]`, though.

DS-Minetest commented on doc/lua_api.txt:
```diff
@@ -2277,16 +2277,37 @@ Elements
+      scrolled.
+    * If this is set to a negative number, the value will be reset to `10`.
+* `largestep=<int>`
+    * Sets scrollbar step value when the bar itself is clicked.
```
`largestep` is no longer used if you click on the tray.
```suggestion
    * Sets scrollbar step value used by page up and page down.
```

v-rob commented:
Rebased (and didn't mess up this time).  Sorry it took so long.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -556,17 +560,30 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
 		spec.send  = true;
-		gui::IGUIScrollBar* e =
-				Environment->addScrollBar(is_horizontal,rect,this,spec.fid);
+		guiScrollBar *e = new guiScrollBar(Environment, this, spec.fid, rect,
```
I see you didn't compile yet.
```suggestion
		GUIScrollBar *e = new GUIScrollBar(Environment, this, spec.fid, rect,
```

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3384,9 +3453,9 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
-						e = static_cast<gui::IGUIScrollBar*>(element);
+					guiScrollBar *e = NULL;
+					if ((element) && (element->getType() == EGUIET_ELEMENT)) {
+						e = static_cast<guiScrollBar *>(element);
```
Here GUI, too.

DS-Minetest commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -27,6 +27,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "inventorymanager.h"
 #include "modalMenu.h"
 #include "guiTable.h"
+#include "guiScrollBar.h"
```
Use forward declaration:
- Move this line to `guiFormSpecMenu.cpp`.
- Add below there (line 36 and further) a line with `class GUIScrollBar;`.

DS-Minetest commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -438,7 +439,7 @@ class GUIFormSpecMenu : public GUIModalMenu
 	std::map<std::string, TooltipSpec> m_tooltips;
 	std::vector<std::pair<irr::core::rect<s32>, TooltipSpec>> m_tooltip_rects;
-	std::vector<std::pair<FieldSpec,gui::IGUIScrollBar*> > m_scrollbars;
+	std::vector<std::pair<FieldSpec, guiScrollBar *>> m_scrollbars;
```
```suggestion
	std::vector<std::pair<FieldSpec, GUIScrollBar *>> m_scrollbars;
```

DS-Minetest commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -542,6 +552,7 @@ class GUIFormSpecMenu : public GUIModalMenu
 	bool parseVersionDirect(const std::string &data);
 	bool parseSizeDirect(parserData* data, const std::string &element);
 	void parseScrollBar(parserData* data, const std::string &element);
+	void parseScrollBarOptions(parserData * data, const std::string &element);
```
```suggestion
	void parseScrollBarOptions(parserData *data, const std::string &element);
```

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -26,6 +26,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "guiButton.h"
 #include "guiFormSpecMenu.h"
 #include "guiTable.h"
+#include "guiScrollBar.h"
```
\*cough\* \*cough\* _…pqr**st**uvwxyz_

v-rob replied:
They're supposed to be in alphabetical order?  I didn't know that.

DS-Minetest commented:
I'd like to point out that this PR is the highest element in a priority list.

v-rob commented:
All this PR needs is reviews, unless people want this integrated into `style`, which would likely be trivial.  But it really should be merged soon, since it's holding everything else up in the queue, and much of those are much more important.

DS-Minetest commented:
https://github.com/minetest/minetest/pull/8740#issuecomment-519513766 also happens in this PR. Here's the image from that comment:
![](https://user-images.githubusercontent.com/2122943/62706364-45bb5680-b9e7-11e9-867d-b1a720fde47f.png)
I request that the visibility of the arrow buttons should be settable with scrollbar options.

DS-Minetest commented:
Another issue is that a scrollbar with `"scrollbaroptions[min=0;max=2;thumbsize=1]"` feels weird when dragging, as if there was some offset. This is probably caused by something in the `GUIScrollBar` class.

v-rob commented:
I'm working on the button visibility, although it's a little trickier than I thought.

> Another issue is that a scrollbar with `"scrollbaroptions[min=0;max=2;thumbsize=1]"` feels weird when dragging, as if there was some offset. This is probably caused by something in the `GUIScrollBar` class.

This completely depends on where you grab the thumb.  If you grab it near the edge, it will move when the mouse approaches the edge of the next position.  Similarly, if you grab it at the center, it will move when the mouse approaches the center of the next position.  So, this is beyond the scope of this PR.

DS-Minetest commented on doc/lua_api.txt:
```diff
@@ -2284,16 +2284,40 @@ Elements
+    * If this is set to zero or less, the value will be reset to `1`.
+* `arrows=<show/hide/default>`
+    * Whether to show the arrow buttons on the scrollbar. `default` hides the arrows
+      when the scrollbar gets too small, but shows them otherwise.
```
The arrow buttons are not hidden if the scrollbar is too small, but if it's too broad. (If it's 4 times wider than long, the buttons are omitted.)

I don't think that the default option is needed. The default can be `show`.

v-rob replied:
The `show` options will _always_ show the arrows, whereas `default` lets the scrollbar choose.  The `show` option can be used, for instance, to only show arrows and no bar, e.g. this code will make a spinbox:
```
scrollbaroptions[arrows=show]
field[0,0;3,1;field;SpinBox;5]
scrollbar[0,3;0.5,1;vertical;sbar;0]
```
If we didn't have `default`, though, then some scrollbars might end up this way accidentally.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -556,23 +561,98 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+			continue;
+		}
+		else if (options[0] == "arrows") {
+			std::string value = trim(options[1]);
```
```suggestion
			const std::string &value = trim(options[1]);
```

SmallJoker replied:
`trim` returns a temporary string, hence const ref is either useless or will even cause issues.

DS-Minetest replied:
Oh, sorry.

DS-Minetest commented on src/gui/guiScrollBar.cpp:
```diff
@@ -315,6 +316,12 @@ void GUIScrollBar::setPageSize(const s32 &size)
+void GUIScrollBar::setArrowsVisible(bool visible)
+{
+	up_button->setVisible(visible);
+	down_button->setVisible(visible);
```
I think, it would be better to add a button visibility attribute (bool or enum).
At least set boarder to width, height or 0.

DS-Minetest commented on src/gui/guiScrollBar.cpp:
```diff
@@ -419,7 +425,12 @@ void GUIScrollBar::refreshControls()
+	if (is_arrows_visible == 2)
+		visible = (border_size != 0);
+	else
+		visible = (bool) is_arrows_visible;
```
Maybe you also have to set `border_size`.

DS-Minetest commented on src/gui/guiScrollBar.h:
```diff
@@ -47,6 +48,7 @@ class GUIScrollBar : public IGUIElement
 
 	IGUIButton *up_button;
 	IGUIButton *down_button;
+	u8 is_arrows_visible;
```
An enum might be better.

rubenwardy commented:
I don't think the properties you state in the OP should be part of style because they're not directly styling or theme related. Most scrollbars should define their own min and max, for example, and having global settings for that doesn't make sense

v-rob commented:
Well, this is a global setting like `tableoptions`, but it can be changed as often as you like.  Global settings make for less code and faster runtime.  Also, if two scrollbars are identical (I can see some possible situations, e.g. RGB color sliders from 0-255), one global setting will work for all of them as long as it's placed right before.

This is no longer WIP, unless anyone has anything else to add/change.

DS-Minetest commented:
There's a weird file called `get_data()`.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -122,6 +123,10 @@ GUIFormSpecMenu::~GUIFormSpecMenu()
 
+	for (auto &scrollbar_it : m_scrollbars) {
+		scrollbar_it.second->drop();
+	}
```
Please also drop in `regenerateGui` (see line 2466).
(And the curly braces aren't needed.)

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -556,23 +561,96 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+		if (is_horizontal)
+			scrollbar_size = dim.X;
+		else
+			scrollbar_size = dim.Y;
```
`s32 scrollbar_size = is_horizontal ? dim.X : dim.Y;`

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -556,23 +561,96 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+	}
+
+	for (const std::string &i : parts) {
+		std::vector<std::string> options = split(i, '=');
```
Would the following be possible?
```suggestion
		const std::vector<std::string> &options = split(i, '=');
```

SmallJoker replied:
No. The function returns a temporary variable.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -556,23 +561,96 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+
+void GUIFormSpecMenu::parseScrollBarOptions(parserData* data, const std::string &element)
+{
+	std::vector<std::string> parts = split(element, ';');
```
Would the following be possible?
```suggestion
	const std::vector<std::string> &parts = split(element, ';');
```

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -556,23 +561,96 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+			data->scrollBarOptions.max = stoi(options[1]);
+			continue;
+		}
+		else if (options[0] == "min") {
```
`}` and `else` in one line:
`} else if (..) {`
(See https://www.kernel.org/doc/html/latest/process/coding-style.html#placing-braces-and-spaces.)

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3384,9 +3467,9 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
-					gui::IGUIScrollBar *e = NULL;
-					if ((element) && (element->getType() == gui::EGUIET_SCROLL_BAR)) {
-						e = static_cast<gui::IGUIScrollBar*>(element);
+					GUIScrollBar *e = NULL;
```
```suggestion
					GUIScrollBar *e = nullptr;
```

DS-Minetest commented:
> Another issue is that a scrollbar with `"scrollbaroptions[min=0;max=2;thumbsize=1]"` feels weird when dragging, as if there was some offset. This is probably caused by something in the `GUIScrollBar` class.
> Edit: This is indeed out of scope, sorry.

Btw. I've found the reason and a fix for this:
`GUIScrollBar::getPosFromMousePos` makes rounding errors. To fix this, change the following line:
https://github.com/minetest/minetest/blob/e892bcdcd7c4144acfe562e16e7add38797fa184/src/gui/guiScrollBar.cpp#L250
to:
`	return core::isnotzero(range()) ? s32(f32(p) / f32(w) * range() + 0.5f) + min_pos : 0;`
(Note the added `0.5f`.)

DS-Minetest commented:
:+1: 
Tested again, works.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -556,23 +559,96 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
 		spec.send  = true;
-		gui::IGUIScrollBar* e =
-				Environment->addScrollBar(is_horizontal,rect,this,spec.fid);
+		GUIScrollBar *e = new GUIScrollBar(Environment, this, spec.fid, rect,
```
Needs `e->drop()`.

DS-Minetest replied:
It is already dropped, see https://github.com/minetest/minetest/pull/8530/files#diff-074850a88796e13dad12dd7ae1a8d9f7R126 and https://github.com/minetest/minetest/pull/8530/files#diff-074850a88796e13dad12dd7ae1a8d9f7R2465.

SmallJoker replied:
Thanks. Although it's currently inconsistent this should work as intended.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -556,23 +559,96 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+			data->scrollBarOptions.max = stoi(options[1]);
+			continue;
+		}
+		else if (options[0] == "min") {
```
Please put `else if` onto the same line as `}` above.

SmallJoker commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -489,6 +490,16 @@ class GUIFormSpecMenu : public GUIModalMenu
+			s32 largeStep = 100;
+			s32 thumbSize = 1;
+			u8 showArrows = 2;
+		} scrollBarOptions;
```
This is a variable name. Please use `snake_case`.

SmallJoker commented on src/gui/guiScrollBar.cpp:
```diff
@@ -419,7 +426,22 @@ void GUIScrollBar::refreshControls()
+		visible = false;
+		border_size = 0;
+	}
+	else {
```
`else` on same line as `}`. Please see coding style guidelines.

SmallJoker commented on src/gui/guiScrollBar.h:
```diff
@@ -47,6 +48,7 @@ class GUIScrollBar : public IGUIElement
 
 	IGUIButton *up_button;
 	IGUIButton *down_button;
+	u8 is_arrows_visible = 2;
```
Why not a boolean?

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -556,23 +559,96 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+		else
+			scrollbar_size = dim.Y;
+
+		e->setPageSize(scrollbar_size * (max-min+1) / data->scrollBarOptions.thumbSize);
```
Spaces between binary operators (`a - b`).

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -556,23 +559,96 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
+		e->setSmallStep(data->scrollBarOptions.smallStep);
+		e->setLargeStep(data->scrollBarOptions.largeStep);
+
+		s32 scrollbar_size;
```
Could be combined to `s32 scrollbar_size = is_horizontal ? dim.X : dim.Y;`

DS-Minetest commented on src/gui/guiScrollBar.h:
```diff
@@ -17,6 +17,12 @@ the arrow buttons where there is insufficient space.
+	HIDE,
+	SHOW,
+	DEFAULT
+};
```
You can put that enum into the `GUIScrollBar` class:
```cpp
class GUIScrollBar : public IGUIElement
{
public:
	enum ArrowVisibility {
		HIDE,
		SHOW,
		DEFAULT
	};
//...
};
```
Outside of this class you can access it with `GUIScrollBar::ArrowVisibility` and its values eg. with `GUIScrollBar::HIDE`.
The forward declaration (`guiFormSpecMenu.h`, line 40)  might be possible with `enum GUIScrollBar::ArrowVisibility;` or with
```cpp
enum GUIScrollBar::ArrowVisibility {
	HIDE,
	SHOW,
	DEFAULT
};
```
(or just include the header in the header).

SmallJoker commented:
@v-rob Due to @DS-Minetest's formspec changes you might need to move some code.

I lost track of this PR, but it apparently only needs re-reviewing after rebase and some quick tests (which I actually wanted to do now).

v-rob commented:
Sorry it took so long to rebase.  I forget how to rebase properly every time I try it, so I wrote it down this time.
